### ISSUE TITLE
QLoRA finetuning of Llama-2 70B not working (GQA mismatch)

### ISSUE THREAD

**Issue by ssmi153** [axolotl-ai-cloud/axolotl#338]
I'm trying to finetune Llama-2 70B on an A100 80GB GPU on Runpod using the Runpod Dockerfile. Using either FlashAttention or Xformers attention I run into the following error: 

For Flashattention:

```
File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/transformers/trainer.py", line 2674, in compute_loss
    outputs = model(**inputs)
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py", line 810, in forward
    outputs = self.model(
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/hooks.py", line 165, in new_forward
    output = old_forward(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py", line 690, in forward
    layer_outputs = torch.utils.checkpoint.checkpoint(
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/utils/checkpoint.py", line 249, in checkpoint
    return CheckpointFunction.apply(function, preserve, *args)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/autograd/function.py", line 506, in apply
    return super().apply(*args, **kwargs)  # type: ignore[misc]
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/utils/checkpoint.py", line 107, in forward
    outputs = run_function(*args)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py", line 686, in custom_forward
    return module(*inputs, output_attentions, None)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/hooks.py", line 165, in new_forward
    output = old_forward(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py", line 413, in forward
    hidden_states, self_attn_weights, present_key_value = self.self_attn(  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/hooks.py", line 165, in new_forward
    output = old_forward(*args, **kwargs)
  File "/workspace/axolotl/src/axolotl/flash_attn.py", line 38, in forward
    .view(bsz, q_len, self.num_heads, self.head_dim)
RuntimeError: shape '[1, 4096, 64, 128]' is invalid for input of size 4194304
```

For Xformers attention:

`[2023-07-30 11:32:17,550] [INFO] [axolotl.scripts.train:219] [PID:1060] loading tokenizer... /external/models/meta-llama_Llama-2-70b-hf loss = self.compute_loss(model, inputs) File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/transformers/trainer.py", line 2671, in compute_loss outputs = model(**inputs) File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/utils/operations.py", line 581, in forward File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py", line 810, in forward outputs = self.model( File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl return forward_call(*args, **kwargs) File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/hooks.py", line 165, in new_forward output = old_forward(*args, **kwargs) File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py", line 690, in forward layer_outputs = torch.utils.checkpoint.checkpoint( File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/utils/checkpoint.py", line 249, in checkpoint return CheckpointFunction.apply(function, preserve, *args) File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/autograd/function.py", line 506, in apply return super().apply(*args, **kwargs)  # type: ignore[misc] File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/utils/checkpoint.py", line 107, in forward outputs = run_function(*args) File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py", line 686, in custom_forward return module(*inputs, output_attentions, None) File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl return forward_call(*args, **kwargs) File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/hooks.py", line 165, in new_forward output = old_forward(*args, **kwargs) File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py", line 413, in forward hidden_states, self_attn_weights, present_key_value = self.self_attn( File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl return forward_call(*args, **kwargs) File "/root/miniconda3/envs/py3.10/lib/python3.10/site-packages/accelerate/hooks.py", line 165, in new_forward output = old_forward(*args, **kwargs) File "/workspace/axolotl/src/axolotl/monkeypatch/llama_attn_hijack_xformers.py", line 48, in xformers_forward .view(bsz, q_len, self.num_heads, self.head_dim) RuntimeError: shape '[1, 4096, 64, 128]' is invalid for input of size 4194304`

This seems to be related to Grouped Query Attention being used in the 70B model.

Take a look at the config.json file for Llama-2 70b: https://huggingface.co/meta-llama/Llama-2-70b-hf/blob/main/config.json
 - num_attention_heads = 64
 - BUT: num_key_value_heads = 8

For the Keys and Values, we're currently multiplying using num_heads (64) when we should be using num_key_value_heads (8).

Looking at the underlying Transformers modeling classes, they've added some more nuance to handle GQA which isn't reflected in our MonkeyPatch file. (https://github.com/huggingface/transformers/commit/07360b6c9c9448d619a82798419ed291dfc6ac8f#diff-06392bad3b9e97be9ade60d4ac46f73b6809388f4d507c2ba1384ab872711c51  - take a look at modeling_llama.py)

I'll have a play around and see if I can patch these changes into our MonkeyPatch file then submit this as a Pull Request if I can get it working.

**Comment by winglian** [comment]
Thanks, this has been on my list. If you have the time, this change should be helpful for adapting the monkeypatch. https://github.com/LAION-AI/Open-Assistant/pull/3595/commits/e36117c53a9777e877094eac9b259632ea9c55de

**Comment by ssmi153** [comment]
Great! I've managed to succesfully update the XFormers monkeypatch to work. I'll submit that as a pull request in a moment once I've finished some more testing. (Testing with the 70B model is SLOW!). I'm running into a whole bunch of other errors with the FlashAttention monkeypatch, but I'll have a look at the commit you linked and see if I can see anything there that might help.

**Comment by MrigankRaman** [comment]
> Great! I've managed to succesfully update the XFormers monkeypatch to work. I'll submit that as a pull request in a moment once I've finished some more testing. (Testing with the 70B model is SLOW!). I'm running into a whole bunch of other errors with the FlashAttention monkeypatch, but I'll have a look at the commit you linked and see if I can see anything there that might help.

Hi can you do it asap. I need it urgently. Thanks a lot

**Comment by ssmi153** [comment]
Here you go: https://github.com/OpenAccess-AI-Collective/axolotl/pull/339  . This works for XFormers Attention. I'm currently running a QLora finetune with LLama-2 70B on an A100 80GB and it seems to be working nicely. If you don't want to wait for the pull request to be merged, you can manually replace your version of `src/axolotl/monkeypatch/llama_attn_hijack_xformers.py` with the updated file in the pull request and it should run without any issues.

I still need to work on the FlashAttention monkeypatch. I'm running into some more issues on that front.

**Comment by MrigankRaman** [comment]
Does it still work with the 13B model? What I wonder is does it have backward compatibility

**Comment by ssmi153** [comment]
Yes, I've tested it on the 13B model too and it's backwards compatible. (If you're just wanting this to work on the 13B model, you don't need to wait for this pull request - it already worked before this. It's just the 70B model that required additional work because it used GQA).

**Comment by MrigankRaman** [comment]
Does it still work with the 13B model? What I wonder is does it have backward compatibility


> Yes, I've tested it on the 13B model too and it's backwards compatible. (If you're just wanting this to work on the 13B model, you don't need to wait for this pull request - it already worked before this. It's just the 70B model that required additional work because it used GQA).

Thanks a lot. Yeah I know. I just need a code that works for both as I am creating a clean codebase. Thanks a lot again

**Comment by winglian** [comment]
@MrigankRaman , @ssmi153's fix has been merged. 

@ssmi153 any interest in updating the regular flash-attn monkey-patch while it's still fresh? 🙏

**Comment by ssmi153** [comment]
@winglian I've tried making the simple changes I made to the XFormers monkeypatch to the FlashAttention monkeypatch, but I then get a different set of errors (FlashAttention only supports F16 and BF16 but it looks like we're possibly passing in Int32 of Float32 datatypes which it doesn't like). Our current FlashAttention monkeypatch is different enough from the other reference implementations that I'm a little bit over my head trying to piece together all the moving parts. I'm probably not going to have time this week to take another look at this, so if someone else wants to have a go in the meantime that's probably a good idea.
